sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/ProactiveOpenConnectionsProcessor.java

```diff
@@ -166,19 +176,24 @@ public void recordOpenConnectionsAndInitCachesStarted(List<CosmosContainerIdenti
                         .getContainerLink(containerIdentity)
                 );
             }
-        }
 
-        if (shouldReInstantiatePublisher) {
-            this.aggressivenessHint.set(ConnectionOpenFlowAggressivenessHint.AGGRESSIVE);
-            this.reInstantiateOpenConnectionsPublisherAndSubscribe(false);
+            if (shouldReInstantiatePublisher) {
+                this.aggressivenessHint.set(ConnectionOpenFlowAggressivenessHint.AGGRESSIVE);
+                this.reInstantiateOpenConnectionsPublisherAndSubscribe(false);
+            }
         }
     }
 
     public Disposable getBackgroundOpenConnectionsPublisher() {
 
         ConcurrencyConfiguration concurrencyConfiguration = concurrencySettings.get(aggressivenessHint.get());
+
         Map<String, List<OpenConnectionTask>> mapSnapshot = new ConcurrentHashMap<>();
-        mapSnapshot.putAll(this.endpointsUnderMonitorMap);
+
+        synchronized (this.endpointsUnderMonitorMapWriteLock) {
```

Comment:
A write lock is acquired here because we wish to have only one thread instantiate the sink and concurrently the read lock can be acquired which will prevent update of `endpointsUnderMonitorMap` and more importantly will prevent task submission to a sink which is still initialized?

Also is there a reason we don't use `endpointsUnderMonitorMapWriteLock.lock()` and `endpointsUnderMonitorMapWriteLock.unlock()`?

Reply:
yes. it will make sure the new tasks will be submitted to the new sink and it also helps to make sure there is no duplicated open connection tasks (tasks has the same addresses) to be submitted

for the lock/unblock, it is a miss, updated